<img src="https://i.ibb.co/K5n24Nd/1.png"  alt="KCode Banner"  />

<a href="https://www.linkedin.com/in/kauebr/"><img src="https://cdn-icons-png.flaticon.com/512/174/174857.png" width="30" ></a> <a href="mailto: kauemandarino@gmail.com"><img src="https://cdn-icons-png.flaticon.com/512/5968/5968534.png" width="30" ></a><a href="
https://api.whatsapp.com/send?phone=5541991835800&text=Oi"> <img src="https://imagepng.org/wp-content/uploads/2017/08/whatsapp-icone-1.png" width="30" ></a><a href="https://github.com/kauebr/"><img src="https://github.githubassets.com/images/modules/logos_page/Octocat.png" width="30" ></a>



In [8]:
from pandas_datareader import data as web
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import date

from calendar import monthrange
from workalendar.america import Brazil
import calendario_dias_uteis as cdu
import matplotlib.pyplot as plt

In [9]:
# Puxando as datas da nossa consulta (2000 ~ 2022)

datas = pd.read_csv('C:/Users/manda/OneDrive/Documents/GitHub/Insights/Dados Siderurgia Brasileira 2016~2022/2 Entrega - Cliclo 2/Passo 05 - Exploração dos dados/Jupyter Notebooks/Dados/UltimosUteis.csv')

In [10]:
datas

Ultimos Uteis
0       1/31/2000
1       2/29/2000
2       3/31/2000
3       4/28/2000
4       5/31/2000
..            ...
268     5/31/2022
269     6/30/2022
270     7/29/2022
271     8/31/2022
272     9/20/2022

[273 rows x 1 columns]

* Constatamos 273 linhas referentes as datas da nossa consulta: ultimos dias úteis de cada mês, de 01/2000 a 08/2022



In [11]:
# Declarando dicionário com as empresas e os códigos na bolsa de valores.

empresas = {'GGBR4.SA': 'Gerdau preferenciais', 'GGBR3.SA': 'Gerdau Metalúrgica preferenciais',
             'VALE3.SA': 'Vale', 'USIM5.SA': 'Usiminas',  'CSNA3.SA': 'CSN'} 
             


* Ampliei o leque de empresas, adcionando a Vale, pois vislubro a possibilidade de inclui-la em minha sugestão de investimento. 

* Agora vou fazer uma consulta no período `macro` para cada empresa, para depois filtrarmos, pois, uma requisição para cada data do nosso dict demoraria cerca de 40 minutos, mas antes vou modelar a estrutura para receber os nossos dados.

In [12]:
# Criando um dicionário raíz chamado cotacoes_acoes para abrigar as listas;
cotacoes_acoes = {} 

# Puxando as cotações
for empresa in empresas:
    cotacao = web.DataReader(empresa, data_source='yahoo', start='01-01-2000', end='08-31-2022')
    cotacoes_acoes[(f"{empresa}")] = cotacao['Adj Close']  
    
cotacoes_acoes



{'GGBR4.SA': Date
 2000-01-03    260.311584
 2000-01-04    254.300262
 2000-01-05    262.995209
 2000-01-06    271.046021
 2000-01-07    283.390656
                  ...    
 2022-08-25     24.670000
 2022-08-26     24.059999
 2022-08-29     23.959999
 2022-08-30     23.540001
 2022-08-31     23.320000
 Name: Adj Close, Length: 5645, dtype: float64,
 'GGBR3.SA': Date
 2000-01-04     0.188095
 2000-01-05     0.188095
 2000-01-06     0.188095
 2000-01-07     0.188095
 2000-01-10     0.188095
                 ...    
 2022-08-25    20.209999
 2022-08-26    19.879999
 2022-08-29    19.780001
 2022-08-30    19.500000
 2022-08-31    19.700001
 Name: Adj Close, Length: 5695, dtype: float64,
 'VALE3.SA': Date
 2000-01-03     1.707494
 2000-01-04     1.666839
 2000-01-05     1.666839
 2000-01-06     1.666839
 2000-01-07     1.727822
                 ...    
 2022-08-25    69.269997
 2022-08-26    68.230003
 2022-08-29    66.910004
 2022-08-30    64.970001
 2022-08-31    64.500000
 Name: Adj Clo

Agora vou transformar as séries em dataframes

In [13]:
for empresa in empresas:
    cotacoes_acoes[empresa] = cotacoes_acoes[empresa].to_frame()
    cotacoes_acoes[empresa].reset_index(inplace=True) # As datas estavam como índice, aqui transformo-as em coluna;

Agora vou juntar todas as cotações e organizar tudo em um único dataframe

In [17]:
cotacao = web.DataReader('GGBR4.SA', data_source='yahoo', start='01-01-2000', end='08-31-2022') # Puxando uma cotação qualquer com o intervalo de datas desejado;
cotacao.reset_index(inplace=True) # As datas estavam como índice, aqui transformo-as em coluna;
cotacao.rename(columns={'Date':'Data'}, inplace=True) # Renomeando a coluna de "Date", para "Data";
df = cotacao['Data'].copy() # Criando uma série, já com as datas;
df.to_csv('C:/Users/manda/OneDrive/Documents/GitHub/Insights/Dados Siderurgia Brasileira 2016~2022/2 Entrega - Cliclo 2/Passo 03 e 04 - Coleta e limpeza dos dados/Dados/datas.csv') # Exportando a série para um arquivo .csv
df = pd.read_csv('C:/Users/manda/OneDrive/Documents/GitHub/Insights/Dados Siderurgia Brasileira 2016~2022/2 Entrega - Cliclo 2/Passo 03 e 04 - Coleta e limpeza dos dados/Dados/datas.csv', usecols=[1]) # Criando um frame de dados a partir do .csv exportado
df

Data
0     2000-01-03
1     2000-01-04
2     2000-01-05
3     2000-01-06
4     2000-01-07
...          ...
5640  2022-08-25
5641  2022-08-26
5642  2022-08-29
5643  2022-08-30
5644  2022-08-31

[5645 rows x 1 columns]

In [32]:
for empresa in empresas:
    df[empresa] = cotacoes_acoes[empresa]['Adj Close']

In [33]:
df

Data    GGBR4.SA   GGBR3.SA   VALE3.SA  USIM5.SA  CSNA3.SA
0     2000-01-03  260.311584   0.188095   1.707494  1.582737  2.079645
1     2000-01-04  254.300262   0.188095   1.666839  1.525033  2.109569
2     2000-01-05  262.995209   0.188095   1.666839  1.670117  2.222848
3     2000-01-06  271.046021   0.188095   1.666839  1.821796  2.323304
4     2000-01-07  283.390656   0.188095   1.727822  1.894338  2.308343
...          ...         ...        ...        ...       ...       ...
5640  2022-08-25   24.670000  19.446829  77.693123  9.370000       NaN
5641  2022-08-26   24.059999  18.276730  77.645584  9.330000       NaN
5642  2022-08-29   23.959999  18.373432  73.595505  8.770000       NaN
5643  2022-08-30   23.540001  18.209040  71.779625  9.010000       NaN
5644  2022-08-31   23.320000  17.715857  72.254982  9.060000       NaN

[5645 rows x 6 columns]